src/server/pps/server/api_server.go

```diff
@@ -1519,19 +1448,197 @@ func (a *apiServer) getLogsFromStats(pachClient *client.APIClient, request *pps.
 					continue
 				}
 
-				mu.Lock()
-				if err := apiGetLogsServer.Send(msg); err != nil {
-					mu.Unlock()
-					return err
+				select {
+				case logCh <- msg:
+				case <-ctx.Done():
+					return nil
 				}
-				mu.Unlock()
 			}
 			return nil
 		})
 	}
 	return eg.Wait()
 }
 
+func (a *apiServer) followLogsForRC(pachClient *client.APIClient, request *pps.GetLogsRequest, containerName, rcName string, logCh chan *pps.LogMessage) error {
+	defer close(logCh)
+	ctx := pachClient.Ctx()
+
+	err := backoff.RetryNotify(func() error {
```

Comment:
This may be a known drawback of this approach, but it occurs to me that a risk is that the `backoff` loop starts, sends a bunch of logs, and then has to retry and sends all the same logs again.

Do you think it would be worth it to inject a log message in the `Notify` fn at the bottom saying something to the effect of "retrying connection, logs after this line may be duplicates"?

Reply:
That's right, and additionally some log lines might be missed (see the PR description.) It's for that reason that I'm hesitant to merge this PR.

Reply:
I see. I re-read the description, and I think I actually don't totally understand the description of when log lines get skipped. Is the idea:
```
----------+-------------
Server    | user session
----------+-------------
Line 1    |  
Line 2    |  
Line 3    |  
Line 4    |  $ pachctl get-logs --tail=3
Line 5    |  Line 2
Line 6    |  Line 3
Line 7    |  Line 4
Line 8    |
Line 9    |  $ pachctl get-logs --tail=3
Line 10   |  Line 7
Line 11   |  Line 8
Line 12   |  Line 9
```

Reply:
`pachctl logs -p <pipeline> --follow` (i.e. follow w/o tail) will repeatedly fetch logs from pods, but may miss logs in the time between requests. However, this will be trivially fixable if I can switch over to getting service logs instead.